### PR TITLE
List.to_string/1 improve error message and docs

### PR DESCRIPTION
**Pull request by eksperimental**
* Make List.to_string/1 error msg non-ambiguous and corrects it mentioning empty lists are valid
* Add elements allowed in List.to_string/1 to @doc
* add tests for empty lists

### CODE REVIEW

File: lib/elixir/lib/list.ex [hunk 1]
```diff
@@ -872,9 +872,9 @@ defmodule List do
 
         To be converted to a string, a list must contain only:
 
-          * strings
-          * integers representing Unicode code points
-          * or a list containing one of these three elements
+          * strings;
+          * integers representing Unicode code points;
+          * and/or a list containing one of these three elements.
```
Comment (whatyouhide) [inline lib/elixir/lib/list.ex]: Could we achieve the same result by removing "or" from what we had here before the change?

Comment (eksperimental) [reply lib/elixir/lib/list.ex]: Nope. because that would be that it must contain all three items

Comment (eksperimental) [reply lib/elixir/lib/list.ex]: We could remove "or" and in the header say "a list must contain either:"

Comment (whatyouhide) [reply lib/elixir/lib/list.ex]: I'm not really sure how so, but then what if we make the first sentence "To be converted to a string, a list must only contain one of the following:"?

Comment (eksperimental) [reply lib/elixir/lib/list.ex]: I like it. let's go with it.

**Comment by eksperimental** [comment]
Ok. I see other issues. a list must be empty and it is valid, and it doesn't follow any of the three items that this error message mention.

File: lib/elixir/lib/list.ex (hunk 1 above)
Comment (eksperimental) [reply lib/elixir/lib/list.ex]: fixed,

Comment (eksperimental) [reply lib/elixir/lib/list.ex]: actually it is wrong. "one of the following" is wrong. it can contain a string and a number

Comment (eksperimental) [reply lib/elixir/lib/list.ex]: I'm changing it for "To be converted to a string, a list must either be empty or only contain the following elements:"

**Comment by eksperimental** [comment]
I have created a second commit that improves the docs and the tests.

**Pull request merged**